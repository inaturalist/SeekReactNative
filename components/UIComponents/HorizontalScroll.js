// @flow

import React, { useRef, useState, useCallback } from "react";
import { Image, FlatList, TouchableOpacity } from "react-native";
import { useRoute } from "@react-navigation/native";
import type { Node } from "react";

import styles from "../../styles/uiComponents/horizontalScroll";
import i18n from "../../i18n";
import icons from "../../assets/icons";
import { dimensions } from "../../styles/global";

type Props = {
  photoList: Array<Object>
}

const HorizontalScroll = ( { photoList }: Props ): Node => {
  const { name } = useRoute();
  const flatList = useRef( null );
  const viewConfigRef = useRef( {
    waitForInteraction: true,
    viewAreaCoveragePercentThreshold: 95
  } );
  const length = photoList.length - 1;
  const { width } = dimensions;
  const [scrollIndex, setScrollIndex] = useState( 0 );

  const isStatsScreen = name === "iNatStats";
  const isSpeciesScreen = name === "Species";
  const isYearInReviewScreen = name === "SeekYearInReview";

  const nextIndex = scrollIndex < length ? scrollIndex + 1 : length;
  const prevIndex = scrollIndex > 0 ? scrollIndex - 1 : 0;

  const scroll = ( index ) => {
    if ( index < 0 || index >= photoList.length ) {
      return;
    }
    if ( flatList && flatList.current !== null ) {
      flatList.current.scrollToIndex( { index, animated: true } );
    }
  };

  const scrollRight = () => scroll( nextIndex );
  const scrollLeft = () => scroll( prevIndex );

  const onViewRef = useRef( ( { changed } ) => {
    const { index } = changed[0];
    if ( !index ) { return; }
    setScrollIndex( index );
  } );

  const renderPhoto = useCallback( ( { item } ) => item, [] );

  // skips measurement of dynamic content for faster loading
  const getItemLayout = useCallback( ( data, index ) => ( {
    length: ( width ),
    offset: ( width ) * index,
    index
  } ), [width] );

  const contentContainerStyle = isYearInReviewScreen ? null : (
    isStatsScreen
      ? styles.photoContainer
      : styles.speciesPhotoContainer
  );

  return (
    <>
      <FlatList
        ref={flatList}
        bounces={false}
        viewabilityConfig={viewConfigRef.current}
        onViewableItemsChanged={onViewRef.current}
        contentContainerStyle={contentContainerStyle}
        data={photoList}
        getItemLayout={getItemLayout}
        horizontal
        indicatorStyle="white"
        initialNumToRender={1}
        pagingEnabled
        renderItem={renderPhoto}
        showsHorizontalScrollIndicator={isStatsScreen}
      />
      {scrollIndex > 0 && (
        <TouchableOpacity
          accessibilityLabel={i18n.t( "accessibility.scroll_left" )}
          accessible
          onPress={scrollLeft}
          style={[
            styles.leftArrow,
            isSpeciesScreen && styles.speciesLeftArrow,
            isYearInReviewScreen && styles.yearInReviewArrow
          ]}
        >
          <Image source={icons.swipeRight} style={styles.rotate} />
        </TouchableOpacity>
      )}
      {scrollIndex < length && (
        <TouchableOpacity
          accessibilityLabel={i18n.t( "accessibility.scroll_right" )}
          accessible
          onPress={scrollRight}
          style={[
            styles.rightArrow,
            isSpeciesScreen && styles.speciesRightArrow,
            isYearInReviewScreen && styles.yearInReviewArrow
          ]}
        >
          <Image source={icons.swipeRight} style={styles.rotateRTL} />
        </TouchableOpacity>
      )}
    </>
  );
};

export default HorizontalScroll;
