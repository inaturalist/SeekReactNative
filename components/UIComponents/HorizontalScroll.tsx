import React, { useRef, useState, useCallback } from "react";
import { View, Image, TouchableOpacity } from "react-native";
import { FlashList, ViewToken } from "@shopify/flash-list";
import { useRoute } from "@react-navigation/native";

import styles from "../../styles/uiComponents/horizontalScroll";
import i18n from "../../i18n";
import icons from "../../assets/icons";

interface Props {
  photoList: JSX.Element[];
}

const HorizontalScroll = ( { photoList }: Props ) => {
  const { name } = useRoute();
  const flashList = useRef<FlashList<JSX.Element>>( null );
  const viewConfigRef = useRef( {
    waitForInteraction: true,
    viewAreaCoveragePercentThreshold: 95
  } );
  const length = photoList.length - 1;
  const [scrollIndex, setScrollIndex] = useState( 0 );

  const isStatsScreen = name === "iNatStats";
  const isSpeciesScreen = name === "Species";
  const isYearInReviewScreen = name === "SeekYearInReview";

  const nextIndex = scrollIndex < length ? scrollIndex + 1 : length;
  const prevIndex = scrollIndex > 0 ? scrollIndex - 1 : 0;

  const scroll = ( index: number ) => {
    if ( index < 0 || index >= photoList.length ) {
      return;
    }
    if ( flashList && flashList.current !== null ) {
      flashList.current?.scrollToIndex( { index, animated: true } );
    }
  };

  const scrollRight = () => scroll( nextIndex );
  const scrollLeft = () => scroll( prevIndex );

  const onViewRef = useRef( ( { changed }: { changed: ViewToken[]} ) => {
    const { index } = changed[0];
    if ( !index ) { return; }
    setScrollIndex( index );
  } );

  const renderPhoto = useCallback( ( { item }: { item: JSX.Element } ) => item, [] );

  const containerStyle = isYearInReviewScreen
    ? null
    : isStatsScreen
    ? styles.bigContainer
    : styles.smallContainer;
  const contentContainerStyle = ( !isYearInReviewScreen && !isStatsScreen ) ? styles.speciesPhotoContainer : {};

  return (
    <View style={containerStyle}>
      <FlashList
        testID="horizontal-scroll"
        ref={flashList}
        bounces={false}
        viewabilityConfig={viewConfigRef.current}
        onViewableItemsChanged={onViewRef.current}
        contentContainerStyle={contentContainerStyle}
        data={photoList}
        horizontal
        indicatorStyle="white"
        pagingEnabled
        renderItem={renderPhoto}
        showsHorizontalScrollIndicator={isStatsScreen}
      />
      {scrollIndex > 0 && (
        <TouchableOpacity
          testID="left-arrow"
          accessibilityLabel={i18n.t( "accessibility.scroll_left" )}
          accessible
          onPress={scrollLeft}
          style={[
            styles.leftArrow,
            isSpeciesScreen && styles.speciesLeftArrow,
            isYearInReviewScreen && styles.yearInReviewArrow
          ]}
        >
          <Image source={icons.swipeRight} style={styles.rotate} />
        </TouchableOpacity>
      )}
      {scrollIndex < length && (
        <TouchableOpacity
          testID="right-arrow"
          accessibilityLabel={i18n.t( "accessibility.scroll_right" )}
          accessible
          onPress={scrollRight}
          style={[
            styles.rightArrow,
            isSpeciesScreen && styles.speciesRightArrow,
            isYearInReviewScreen && styles.yearInReviewArrow
          ]}
        >
          <Image source={icons.swipeRight} style={styles.rotateRTL} />
        </TouchableOpacity>
      )}
    </View>
  );
};

export default HorizontalScroll;
