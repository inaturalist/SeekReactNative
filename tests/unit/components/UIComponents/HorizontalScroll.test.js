import React from "react";
import { View, Text } from "react-native";
import { render, screen } from "tests/jest-utils";

import HorizontalScroll from "../../../../components/UIComponents/HorizontalScroll";

const mockPhotoList = [
    <View key={0} >
        <Text>Item 1</Text>
    </View>,
    <View key={1} >
        <Text>Item 2</Text>
    </View>,
    <View key={2} >
        <Text>Item 3</Text>
    </View>
];

const containerTestId = "horizontal-scroll";

describe( "HorizontalScroll", () => {
  test( "should render correctly", async () => {
    render( <HorizontalScroll photoList={mockPhotoList} /> );
    await screen.findByTestId( containerTestId );
    expect( screen ).toMatchSnapshot();
  } );

  test( "should display the first item in props", async () => {
    render( <HorizontalScroll photoList={mockPhotoList} /> );
    const item1 = await screen.findByText( "Item 1" );
    expect( item1 ).toBeTruthy();
  } );
} );
