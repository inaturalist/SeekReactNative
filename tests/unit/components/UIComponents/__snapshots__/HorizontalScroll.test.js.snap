// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalScroll should render correctly 1`] = `
<View
  style={
    {
      "height": 250,
    }
  }
>
  <View
    style={{}}
  >
    <RCTScrollView
      applyWindowCorrection={[Function]}
      bounces={false}
      canChangeSize={true}
      contentContainerStyle={
        {
          "backgroundColor": "#000000",
          "minHeight": 1,
          "minWidth": 1,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
      contentHeight={900}
      contentWidth={1050}
      data={
        [
          <View>
            <Text>
              Item 1
            </Text>
          </View>,
          <View>
            <Text>
              Item 2
            </Text>
          </View>,
          <View>
            <Text>
              Item 3
            </Text>
          </View>,
        ]
      }
      dataProvider={
        DataProvider {
          "_data": [
            <View>
              <Text>
                Item 1
              </Text>
            </View>,
            <View>
              <Text>
                Item 2
              </Text>
            </View>,
            <View>
              <Text>
                Item 3
              </Text>
            </View>,
          ],
          "_firstIndexToProcess": 0,
          "_hasStableIds": false,
          "_requiresDataChangeHandling": false,
          "_size": 3,
          "getStableId": [Function],
          "rowHasChanged": [Function],
        }
      }
      disableRecycling={false}
      estimatedItemSize={350}
      extendedState={{}}
      externalScrollView={[Function]}
      finalRenderAheadOffset={250}
      forceNonDeterministicRendering={true}
      getItemLayout={[Function]}
      horizontal={true}
      indicatorStyle="white"
      initialNumToRender={1}
      initialOffset={0}
      initialRenderIndex={0}
      isHorizontal={true}
      layoutProvider={
        GridLayoutProviderWithProps {
          "_acceptableRelayoutDelta": 1,
          "_getHeightOrWidth": [Function],
          "_getLayoutTypeForIndex": [Function],
          "_getSpan": [Function],
          "_hasExpired": false,
          "_isHorizontal": true,
          "_lastLayoutManager": GridLayoutManager {
            "_acceptableRelayoutDelta": 1,
            "_getSpan": [Function],
            "_isGridHorizontal": true,
            "_isHorizontal": true,
            "_layoutProvider": [Circular],
            "_layouts": [
              {
                "height": 900,
                "type": 0,
                "width": 350,
                "x": 0,
                "y": 0,
              },
              {
                "height": 900,
                "type": 0,
                "width": 350,
                "x": 350,
                "y": 0,
              },
              {
                "height": 900,
                "type": 0,
                "width": 350,
                "x": 700,
                "y": 0,
              },
            ],
            "_maxSpan": 1,
            "_renderWindowSize": {
              "height": 900,
              "width": 400,
            },
            "_totalHeight": 900,
            "_totalWidth": 1050,
            "_window": {
              "height": 900,
              "width": 400,
            },
          },
          "_maxSpan": 1,
          "_renderWindowSize": {
            "height": 900,
            "width": 400,
          },
          "_setLayoutForType": [Function],
          "_tempDim": {
            "height": 0,
            "width": 0,
          },
          "averageWindow": AverageWindow {
            "currentAverage": 350,
            "currentCount": 1,
            "inputValues": [
              350,
              ,
              ,
              ,
              ,
              ,
            ],
            "nextIndex": 1,
          },
          "defaultEstimatedItemSize": 100,
          "layoutObject": {
            "size": undefined,
            "span": undefined,
          },
          "props": {
            "bounces": false,
            "contentContainerStyle": {
              "backgroundColor": "#000000",
            },
            "data": [
              <View>
                <Text>
                  Item 1
                </Text>
              </View>,
              <View>
                <Text>
                  Item 2
                </Text>
              </View>,
              <View>
                <Text>
                  Item 3
                </Text>
              </View>,
            ],
            "estimatedItemSize": 350,
            "getItemLayout": [Function],
            "horizontal": true,
            "indicatorStyle": "white",
            "initialNumToRender": 1,
            "numColumns": 1,
            "onViewableItemsChanged": [Function],
            "pagingEnabled": true,
            "renderItem": [Function],
            "showsHorizontalScrollIndicator": false,
            "testID": "horizontal-scroll",
            "viewabilityConfig": {
              "viewAreaCoveragePercentThreshold": 95,
              "waitForInteraction": true,
            },
          },
          "renderWindowInsets": {
            "height": 0,
            "width": 0,
          },
          "shouldRefreshWithAnchoring": true,
        }
      }
      maxRenderAhead={750}
      numColumns={1}
      onEndReached={[Function]}
      onEndReachedThreshold={0}
      onEndReachedThresholdRelative={0}
      onItemLayout={[Function]}
      onLayout={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onSizeChanged={[Function]}
      onViewableItemsChanged={[Function]}
      onVisibleIndicesChanged={[Function]}
      pagingEnabled={true}
      removeClippedSubviews={false}
      renderAheadOffset={0}
      renderAheadStep={250}
      renderContentContainer={[Function]}
      renderItem={[Function]}
      renderItemContainer={[Function]}
      rowRenderer={[Function]}
      scrollEventThrottle={16}
      scrollThrottle={16}
      scrollViewProps={
        {
          "contentContainerStyle": {
            "backgroundColor": "#000000",
            "minHeight": 1,
            "minWidth": 1,
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          "onLayout": [Function],
          "onScrollBeginDrag": [Function],
          "refreshControl": undefined,
          "style": {
            "minHeight": 1,
            "minWidth": 1,
          },
        }
      }
      showsHorizontalScrollIndicator={false}
      style={
        {
          "minHeight": 1,
          "minWidth": 1,
        }
      }
      suppressBoundedSizeException={true}
      testID="horizontal-scroll"
      viewabilityConfig={
        {
          "viewAreaCoveragePercentThreshold": 95,
          "waitForInteraction": true,
        }
      }
      windowCorrectionConfig={
        {
          "applyToInitialOffset": true,
          "applyToItemScroll": true,
          "value": {
            "endCorrection": 0,
            "startCorrection": 0,
            "windowShift": -0,
          },
        }
      }
    >
      <View>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "paddingLeft": 0,
                "paddingTop": 0,
              }
            }
          />
          <View
            style={
              [
                undefined,
                undefined,
              ]
            }
          />
          <AutoLayoutView
            enableInstrumentation={false}
            horizontal={true}
            onBlankAreaEvent={[Function]}
            onLayout={[Function]}
            renderAheadOffset={0}
            scrollOffset={0}
            style={
              {
                "height": 900,
                "width": 1050,
              }
            }
            windowSize={400}
          >
            <CellContainer
              index={0}
              onLayout={[Function]}
              style={
                {
                  "alignItems": "stretch",
                  "flexDirection": "row",
                  "height": 900,
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "column",
                  }
                }
              >
                <View>
                  <Text>
                    Item 1
                  </Text>
                </View>
              </View>
            </CellContainer>
            <CellContainer
              index={1}
              onLayout={[Function]}
              style={
                {
                  "alignItems": "stretch",
                  "flexDirection": "row",
                  "height": 900,
                  "left": 350,
                  "position": "absolute",
                  "top": 0,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "column",
                  }
                }
              >
                <View>
                  <Text>
                    Item 2
                  </Text>
                </View>
              </View>
            </CellContainer>
          </AutoLayoutView>
          <CellContainer
            index={-1}
            style={
              [
                undefined,
                undefined,
              ]
            }
          />
          <View
            style={
              {
                "paddingBottom": 0,
                "paddingRight": 0,
              }
            }
          />
          <View
            pointerEvents="none"
            style={
              {
                "opacity": 0,
              }
            }
          >
            <View>
              <Text>
                Item 2
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    accessibilityLabel="Scroll right"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "paddingLeft": 20,
        "paddingRight": 5,
        "paddingVertical": 20,
        "position": "absolute",
        "right": 0,
        "top": 1010,
        "zIndex": 1,
      }
    }
    testID="right-arrow"
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons/img-swipearrow-right.webp",
        }
      }
      style={
        {
          "transform": [
            {
              "rotate": "0deg",
            },
          ],
        }
      }
    />
  </View>
</View>
`;
